test/parachain/tests/integration_test.rs

```diff
@@ -0,0 +1,390 @@
+#![allow(unused_variables, dead_code)]
+
+use assert_cmd::cargo::cargo_bin;
+use async_std::task::sleep;
+use codec::Encode;
+use futures::{future::FutureExt, pin_mut, select};
+use polkadot_primitives::parachain::{Info, Scheduling};
+use polkadot_primitives::Hash as PHash;
+use polkadot_runtime::{Header, OnlyStakingAndClaims, Runtime, SignedExtra, SignedPayload};
+use polkadot_runtime_common::{parachains, registrar, BlockHashCount};
+use serde_json::Value;
+use sp_arithmetic::traits::SaturatedConversion;
+use sp_runtime::generic;
+use sp_version::RuntimeVersion;
+use std::collections::HashSet;
+use std::{
+	convert::TryInto,
+	env, fs,
+	io::Read,
+	net,
+	path::PathBuf,
+	process::{Child, Command, Stdio},
+	thread,
+	time::Duration,
+};
+use substrate_test_runtime_client::AccountKeyring::Alice;
+use tempfile::tempdir;
+
+static POLKADOT_ARGS: &[&str] = &["polkadot", "--chain=res/polkadot_chainspec.json"];
+
+jsonrpsee::rpc_api! {
+	Author {
+		#[rpc(method = "author_submitExtrinsic", positional_params)]
+		fn submit_extrinsic(extrinsic: String) -> PHash;
+	}
+
+	Chain {
+		#[rpc(method = "chain_getFinalizedHead")]
+		fn current_block_hash() -> PHash;
+
+		#[rpc(method = "chain_getHeader", positional_params)]
+		fn header(hash: PHash) -> Option<Header>;
+
+		#[rpc(method = "chain_getBlockHash", positional_params)]
+		fn block_hash(hash: Option<u64>) -> Option<PHash>;
+	}
+
+	State {
+		#[rpc(method = "state_getRuntimeVersion")]
+		fn runtime_version() -> RuntimeVersion;
+	}
+
+	System {
+		#[rpc(method = "system_networkState")]
+		fn network_state() -> Value;
+	}
+}
+
+// Adapted from
+// https://github.com/rust-lang/cargo/blob/485670b3983b52289a2f353d589c57fae2f60f82/tests/testsuite/support/mod.rs#L507
+fn target_dir() -> PathBuf {
+	env::current_exe()
+		.ok()
+		.map(|mut path| {
+			path.pop();
+			if path.ends_with("deps") {
+				path.pop();
+			}
+			path
+		})
+		.unwrap()
+}
+
+struct ChildHelper<'a> {
+	name: String,
+	child: &'a mut Child,
+}
+
+impl<'a> Drop for ChildHelper<'a> {
+	fn drop(&mut self) {
+		let name = self.name.clone();
+
+		self.terminate();
+
+		let mut stdout = String::new();
+		if let Some(reader) = self.child.stdout.as_mut() {
+			let _ = reader.read_to_string(&mut stdout);
+		}
+		eprintln!("process '{}' stdout:\n{}\n", name, stdout,);
+
+		let mut stderr = String::new();
+		if let Some(reader) = self.child.stderr.as_mut() {
+			let _ = reader.read_to_string(&mut stderr);
+		}
+		eprintln!("process '{}' stderr:\n{}\n", name, stderr,);
+	}
+}
+
+impl<'a> ChildHelper<'a> {
+	fn new(name: &str, child: &'a mut Child) -> ChildHelper<'a> {
+		ChildHelper {
+			name: name.to_string(),
+			child,
+		}
+	}
+
+	fn terminate(&mut self) {
+		match self.child.try_wait() {
+			Ok(Some(_)) => return,
+			Ok(None) => {}
+			Err(err) => {
+				eprintln!("could not wait for child process to finish: {}", err);
+				let _ = self.child.kill();
+				let _ = self.child.wait();
+				return;
+			}
+		}
+
+		#[cfg(unix)]
```

Comment:
The default kill method does really send SIGKILL to the process. This one will attempt to terminate gracefully if it can... I'm not sure if that's absolutely necessary but it's less surprising: if I want to restart a node using the existing config dir or if I want to check what's inside after exit...

Reply:
If it can not end it gracefully, there is a bug.

Reply:
I think we should remove this extra code here.